### DISCUSSION TITLE
Run swarm_cse_pt job with brats-learner

### DISCUSSION THREAD

**Discussion by jingyunc** [NVIDIA/NVFlare#2113] [Q&A] [answered]
I'm trying to create a swarm_cse_pt job with brats data/learner. When directly replacing the cifar10-learner (in Swarm Learning example) with brats-learner (in Brats18 example), I got an error:
`CrossSiteEvalClientController[638795] ERROR [identity=site-1, run=simulate_job, peer=site-3, peer_run=simulate_job, task_name=cse_ask_for_model]: failed to get local model from submit_model_executor: TASK_UNSUPPORTED`

I noticed in Swarm Learning example (job_templates/swarm_cse_pt/config_fed_client.json), **ModelLearnerExecutor** was used for cifar10-learner: 
```
"executor": {
        "id": "Executor",
        "path": "nvflare.app_common.executors.model_learner_executor.ModelLearnerExecutor",
        "args": {
          "learner_id": "cifar10-learner"
        }
```

while in Brats18 example (examples/advanced/brats18/configs/brats_fedavg/app/config/config_fed_client.json),  **LearnerExecutor** was used for brats-learner:
```
 "executor": {
        "id": "Executor",
        "name": "LearnerExecutor",
        "args": {
          "learner_id": "brats-learner"
        }
```

But when I tried to pair **ModelLeanerExecutor** with brats-learner in my configuration, I got following error:
`TypeError: learner must be <class 'nvflare.app_common.abstract.model_learner.ModelLearner'>, but got <class 'pt.learners.supervised_monai_brats_learner.SupervisedMonaiBratsLearner'>.
`

Is there any way I can modify the brats-learner (supervised_monai_brats_learner) so that it can be paired with ModelLearnerExecutor (i.e. support submit_model task), or maybe there is a "ModelLearner version" of brats-learner already predefined somewhere? Thanks.

**Comment by YuanTingHsieh** [comment]
@jingyunc as of now, we have only one example of ModelLearner: https://github.com/NVIDIA/NVFlare/blob/main/examples/advanced/cifar10/pt/learners/cifar10_model_learner.py

Feel free to write your own ModelLearner for brats.

Another thing you could do is switch the ModelLearnerExecutor to LearnerExecutor, then you don't need to write your own ModelLearner for brats.

LearnerExectuor also support submit_model task: https://github.com/NVIDIA/NVFlare/blob/main/nvflare/app_common/executors/learner_executor.py#L121

**Reply by SYangster** [reply, marked as answer]
Adding to above, if choosing to switch to use` LearnerExecutor` and modify the current `SupervisedMonaiBratsLearner` to support `submit_model task`, you would need to implement `get_model_for_validation()` which retrieves the best local model saved from training. 

Currently, the model is not being saved locally, but you can reference https://github.com/NVIDIA/NVFlare/blob/main/examples/advanced/cifar10/pt/learners/cifar10_learner.py#L311 to see how this logic can be done in a Learner.
